hooks/platform.go

```diff
@@ -0,0 +1,158 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// Package hooks provides a way to protect keys using hooks supplied via a gadget
+// and kernel snap for a device.
+//
+// It provides a mechanism to bind model assertions and boot modes to key data objects.
+// As enforcing of the binding of these is performed in software, this platform relies on
+// another mechanism to secure the environment in which keys are recovered, such as verified
+// boot which is non configurable or not configurable in a way that would allow execution of
+// components that do not enforce these binding.
+package hooks
+
+import (
+	"crypto/aes"
+	"crypto/cipher"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/snapcore/secboot"
+)
+
+const (
+	platformName = "fde-hooks-v3"
+)
+
+var (
+	keyRevealerMu sync.Mutex
+	keyRevealer   KeyRevealer = nullKeyRevealer{}
+)
+
+type hooksPlatform struct{}
+
+func (*hooksPlatform) RecoverKeys(data *secboot.PlatformKeyData, encryptedPayload []byte) ([]byte, error) {
+	var kd KeyData
+	if err := json.Unmarshal(data.EncodedHandle, &kd); err != nil {
+		return nil, &secboot.PlatformHandlerError{
+			Type: secboot.PlatformHandlerErrorInvalidData,
+			Err:  err,
+		}
+	}
+
+	if err := kd.data.Scope.IsBootEnvironmentAuthorized(); err != nil {
+		return nil, &secboot.PlatformHandlerError{
+			Type: secboot.PlatformHandlerErrorInvalidData,
+			Err:  fmt.Errorf("cannot authorize boot environment: %w", err),
+		}
+	}
+
+	aad, err := kd.data.Scope.MakeAEADAdditionalData(data.Generation, data.KDFAlg, data.AuthMode)
+	if err != nil {
+		return nil, &secboot.PlatformHandlerError{
+			Type: secboot.PlatformHandlerErrorInvalidData,
+			Err:  fmt.Errorf("cannot make AAD: %w", err),
+		}
+	}
+
+	keyRevealerMu.Lock()
+	defer keyRevealerMu.Unlock()
+
+	switch {
+	case kd.data.AEADCompat != nil:
+		symKey, err := keyRevealer.RevealKey(kd.data.Handle, kd.data.AEADCompat.EncryptedKey, nil)
+		if err != nil {
```

Comment:
same for the revealer

Reply:
I've added tests for key revealer errors, plus tests for several other error paths.